lang/src/org/partiql/lang/eval/builtins/ExtractExprFunction.kt

```diff
@@ -38,22 +39,39 @@ internal class ExtractExprFunction(valueFactory: ExprValueFactory) : NullPropaga
 
     private fun Timestamp.minuteOffset() = (localOffset ?: 0) % SECONDS_PER_MINUTE
 
+    private fun extractedValue(datePart: DatePart, dateTimeValue: Any?) : Int {
+        return when (dateTimeValue) {
+            is Timestamp -> when (datePart) {
+                DatePart.YEAR -> dateTimeValue.year
+                DatePart.MONTH -> dateTimeValue.month
+                DatePart.DAY -> dateTimeValue.day
+                DatePart.HOUR -> dateTimeValue.hour
+                DatePart.MINUTE -> dateTimeValue.minute
+                DatePart.SECOND -> dateTimeValue.second
+                DatePart.TIMEZONE_HOUR -> dateTimeValue.hourOffset()
+                DatePart.TIMEZONE_MINUTE -> dateTimeValue.minuteOffset()
+            }
+            is LocalDate -> when (datePart) {
+                DatePart.YEAR -> dateTimeValue.year
+                DatePart.MONTH -> dateTimeValue.monthValue
+                DatePart.DAY -> dateTimeValue.dayOfMonth
+                DatePart.TIMEZONE_HOUR,
+                DatePart.TIMEZONE_MINUTE -> errNoContext("Timestamp unit ${datePart.name.toLowerCase()} not supported for DATE type", internal = false)
+                else -> 0
```

Comment:
Is `0` the right return value here or should accessing the other date components be an error - or missing in permissive mode?

I'm not sure I know the answer with this, please follow up in the team channel.  

We say that there is no time component for the `DATE` type i.e. it represents a specific day, not midnight on a specific day, but returning zero for the other fields would seem to be inconsistent with that.

Also, in general it is better to avoid the use of `else` in your `when` expression because for the uninitiated reader, it's not easy to tell if your intent for the other `DatePart` values was return zero or if a branch for a specific `DatePart` was inadvertently omitted.  Also, if you don't use `else` and then another `DatePart` is later added, the compiler will be helpful to inform you that you missed a branch here.

Reply:
So I checked with PostgreSQL to confirm that it returns zero for `hour`, `minute`,  `second`. It does throw an error for `timezone_hour` or `timezone_minute` as those are invalid for `date`. We also return `0` from `timestamp` but in that case those are the default values. But considering we do return `0` from `extract(hour from ``2007-01-01T``)`, I think returning `0` is being consistent.

Changed `else` to include the remaining branches in `when`.